ffi/derive/src/impl_visitor.rs

```diff
@@ -261,34 +262,39 @@ impl<'ast> Visit<'ast> for ImplVisitor<'ast> {
         if node.unsafety.is_some() {
             // NOTE: Its's irrelevant
         }
-        self.trait_name = node.trait_.as_ref().map(|trait_| &trait_.1);
+        self.trait_name = node.trait_.as_ref().map(|(_, trait_, _)| trait_);
         self.visit_self_type(&*node.self_ty);
 
-        for it in &node.items {
-            match it {
+        let self_ty = self.self_ty.expect_or_abort("Defined");
+        self.fns
+            .extend(node.items.iter().filter_map(|item| match item {
                 syn::ImplItem::Method(method) => {
-                    let self_ty = self.self_ty.expect_or_abort("Defined");
-                    self.fns
-                        .push(FnDescriptor::from_impl_method(self_ty, method))
-                }
-                syn::ImplItem::Type(type_) => {
-                    self.associated_types.push((&type_.ident, &type_.ty));
+                    // NOTE: private methods in inherent impl are skipped
+                    if self.trait_name.is_none()
+                        && !matches!(method.vis, syn::Visibility::Public(_))
+                    {
+                        return None;
+                    }
+                    Some(FnDescriptor::from_impl_method(
+                        self_ty,
+                        self.trait_name,
+                        method,
+                    ))
                 }
-                _ => abort!(
-                    node,
-                    "Only methods or types are supported inside impl blocks"
-                ),
-            }
-        }
+                _ => None,
+            }));
     }
 }
 
 impl<'ast> Visit<'ast> for FnVisitor<'ast> {
     fn visit_impl_item_method(&mut self, node: &'ast syn::ImplItemMethod) {
         self.doc = find_doc_attr(&node.attrs).cloned();
 
-        if !matches!(node.vis, syn::Visibility::Public(_)) {
-            abort!(node.vis, "Methods defined in the impl block must be public");
+        if self.trait_name.is_none() && !matches!(node.vis, syn::Visibility::Public(_)) {
+            abort!(
+                node.vis,
+                "Private methods defined in an inherent `impl` block should not be exported, this is a bug in the library",
+            );
```

Comment:
shouldn't this be removed now?

Reply:
or is this a freestanding function and therefore it makes no sense to attach `ffi_export` on it which this message is about